lib/Runtime/Library/JavascriptObject.cpp

```diff
@@ -2060,4 +2060,195 @@ namespace Js
         }
         return returnValue;
     }
+
+    /*static*/
+    void JavascriptObject::CopyDataPropertiesHelper(Var source, RecyclableObject* to, PropertyId* excluded, uint32 excludedLength, ScriptContext* scriptContext)
+    {
+        // 1. Assert Type(to) is Object.
+        // 2. Assert Type(excluded) is List.
+        // 3. If source is undefined or null, let keys be a new empty List
+        // 4. Else,
+        //        a. Let from be ToObject(source).
+        //        b. Let keys be from.[[OwnpropertyKeys]]().
+        // 5. Repeat for each element nextKey of keys in List order,
+        //        a. Let found be false.
+        //        b. Repeat for each element e of excluded,
+        //            i. If e is not empty and SameValue(e, nextKey) is true, then
+        //                1. Set found to true.
+        //        c. If found is false, then
+        //            i. Let desc be from.[[GetOwnProperty]](nextKey).
+        //            ii. If desc is not undefined and desc.[[Enumerable]] is true, then
+        //                1. Let propValue be Get(from, nextKey).
+        //                2. Perform CreateDataProperty(to, nextKey, propValue).
+        // 6. Return target (target is returned in pointer).
+        Assert(JavascriptOperators::IsObject(to));
+        RecyclableObject* from = nullptr;
+        if (!JavascriptConversion::ToObject(source, scriptContext, &from))
+        {
+            if (JavascriptOperators::IsUndefinedOrNull(source))
+            {
+                return;
+            }
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedObject, _u("Object.CopyDataPropertiesHelper"));
+        }
+
+#if ENABLE_COPYONACCESS_ARRAY
+        JavascriptLibrary::CheckAndConvertCopyOnAccessNativeIntArray<Var>(from);
+#endif
+        // if proxy, take slow path by calling [[OwnPropertyKeys]] on source
+        if (JavascriptProxy::Is(from))
+        {
+            CopyDataPropertiesForProxyObjects(from, to, excluded, excludedLength, scriptContext);
+        }
+        // else use enumerator to extract keys from source
+        else
+        {
+            CopyDataPropertiesForGenericObjects(from, to, excluded, excludedLength, scriptContext);
+        }
+    }
+
+    /*static*/
+    void JavascriptObject::CopyDataPropertiesForGenericObjects(RecyclableObject* from, RecyclableObject* to, PropertyId* excluded, uint32 excludedLength, ScriptContext* scriptContext)
```

Comment:
Is there any way we could reduce the code duplication between this and Object.assign? For example, maybe make the function a template `template <bool callSetters, bool checkExclusionList>` and then use `if` blocks to control which pieces run? I'm not claiming to have the only or best solution, but I think we could make the code shorter to read and easier to maintain.